mysys/checksum.c

```diff
@@ -18,6 +18,48 @@
 #include <my_sys.h>
 #include <zlib.h>
 
+/* TODO: remove this once zlib adds inherent support for hardware accelerated
+crc32 for all architectures. */
+typedef unsigned long (*my_crc32_func_t)(unsigned long crc,
+                                         const unsigned char *ptr,
+                                         unsigned int len);
+static my_crc32_func_t my_crc32= crc32;
+
+#if __GNUC__ >= 4 && defined(__x86_64__) && defined(HAVE_CLMUL_INSTRUCTION)
+/*----------------------------- x86_64 ---------------------------------*/
+extern int is_pclmul_enabled();
+extern unsigned long crc32_pclmul(unsigned long crc32,
+                                  const unsigned char *buf, unsigned int len);
+void my_crc32_init()
+{
+  my_crc32= is_pclmul_enabled() ? crc32_pclmul : crc32;
+}
```

Comment:
Could we define the function `my_crc32_init()` as part of the special implementations, so that we avoid exporting two symbols per implementation: ("is enabled" and the actual implementation)?

I think that the globally defined function pointer should be called `my_checksum()`, and the initialization function could be called `my_checksum_init()` for consistency.

If the zlib `crc32()` function has such an unfortunate call interface, then I think we’d better wrap it, and define a cleaner interface, something like the following:
```c
typedef unsigned (*my_checksum_t)(unsigned, const void *, size_t);

static unsigned my_crc32_zlib(unsigned crc, const void *data, size_t len)
{
  return (unsigned) crc32(crc, data, (unsigned) len);
}

my_checksum_t my_checksum= my_crc32_zlib;

__attribute__((weak)) void my_checksum_init(void) {}
```
Please remove the `defined(__x86_64__)`. It seems to me that `HAVE_CLMUL_INSTRUCTION` alone is sufficient.

Reply:
- my_crc32_init ... moving it as part of the special implementation.
we can do it that but then we will have to export the my_crc32_func_t so either way we have to export something. So I reserved the decision and thought maybe it is better to export it in checksum that polluting the clear crc32 special implementation.

- my_checkum vs my_crc32 .. my_checksum also does the dbug-print (not sure how useful it is but act as a good old point to debug things). From run-time perspective anyway it is inline so not leading to any overhead. also checksum is logical concept that for which we have used crc32 if in future we plan to use other algorithm this indirection could help keep the code clean.

- I have integrated a new my_crc32_zlib wrapper and adjusted existing x86 and arm wrapper to match it.